libs/geometry/src/TangentSpaceMesh.cpp

```diff
@@ -255,16 +353,47 @@ Builder& Builder::algorithm(Algorithm algo) noexcept {
 }
 
 TangentSpaceMesh* Builder::build() {
-    mMesh->mOutput->algorithm = selectAlgorithm(mMesh->mInput);
-    // "Working in progress. Not for use"
-    return NULL;
+    // Work in progress. Not for use.
+    Algorithm algo = selectAlgorithm(mMesh->mInput);
+    MethodPtr method = nullptr;
+    switch (algo) {
+        case Algorithm::FRISVAD:
+            method = [](const TangentSpaceMeshInput* input, TangentSpaceMeshOutput* output) {
+                normalsOnlyMethod(input, output, frisvadKernel);
+            };
+            break;
+        case Algorithm::HUGHES_MOLLER:
+            method = [](const TangentSpaceMeshInput* input, TangentSpaceMeshOutput* output) {
+                normalsOnlyMethod(input, output, hughesMollerKernel);
+            };
+            break;
+        case Algorithm::FLAT_SHADING:
+            method = flatShadingMethod;
+            break;
+        default:
+            break;
+    }
+    assert_invariant(method);
+    method(mMesh->mInput, mMesh->mOutput);
+    return mMesh;
 }
 
 TangentSpaceMesh::TangentSpaceMesh() noexcept
         :mInput(new TangentSpaceMeshInput()), mOutput(new TangentSpaceMeshOutput()) {
 }
 
-TangentSpaceMesh::~TangentSpaceMesh() noexcept {}
+#define CLEAN_HELPER(ptr, inputPtr) \
+   if (ptr && ptr != inputPtr) delete [] ptr;\
+   ptr = nullptr
+
+TangentSpaceMesh::~TangentSpaceMesh() noexcept {
+    CLEAN_HELPER(mOutput->tangentSpace, nullptr);
+    CLEAN_HELPER(mOutput->uvs, mInput->uvs);
+    CLEAN_HELPER(mOutput->positions, mInput->positions);
+    CLEAN_HELPER(mOutput->triangles16, mInput->triangles16);
+    CLEAN_HELPER(mOutput->triangles32, mInput->triangles32);
```

Comment:
This makes me a little nervous. It's probably fine for now, but maybe consider using smart pointers or creating separate pointers in the `mOutput` structure to clearly demarcate between user-provided data and data this class is responsible for freeing.

Reply:
noted. Will switch to smart pointers in the next patch.